### ISSUE TITLE
`notifyPropertyChange` does not propagate from controller to component

### ISSUE THREAD

**Issue by lolmaus** [emberjs/ember.js#10661]
Demo: http://emberjs.jsbin.com/xufitu/3/edit?html,js,console,output

```handlebars
    {{foo-bar myProperty=myProperty}}
    <button {{action "trigger"}}>Trigger</button>
```

```js
App = Ember.Application.create();

App.IndexController = Ember.Controller.extend({
  myProperty: { zomg: 'lol' },
  
  report: function() {
    Em.debug('ConTROLLer: property did change!');
  }.observes('myProperty'),
  
  actions: {
    trigger: function() {
      this.notifyPropertyChange('myProperty');
    }
  }
});


App.FooBarComponent = Ember.Component.extend({  
  myProperty: null, 
  
  report: function() {
    Em.debug('ComPONent: property did change!');
  }.observes('myProperty')
});
```

When i trigger the controller action, i expect both observers to fire.

In reality, only the controller's observer fires and the component's does not.

![dafuq](http://i.imgur.com/nJ2rGJ5.jpg)

**Comment by lolmaus** [comment]
Probably a duplicate of #4487, but that one is closed with an irrelevant proposed solution.

**Comment by OrKoN** [comment]
@lolmaus there is a workaround http://stackoverflow.com/a/25523850/601499 But it would great if this worked.

**Comment by alexspeller** [comment]
This really feels like it should work. Perhaps this is intentional but it seems weird, and if so it should probably be called out explicitly in the docs that it doesn't work.

**Comment by lolmaus** [comment]
@OrKoN, thank you for your suggestion. Like the author of that question noted, it's a workaround and it smells boilerplate code. Both him and i believe it should be addressed.

**Comment by lolmaus** [comment]
@grapho has discovered that this works normally for computed properties: http://emberjs.jsbin.com/xufitu/4/edit?html,js,console,output

So only normal properties are affected.

**Comment by stefanpenner** [comment]
it appears nothing in your component consumes the CP, so it is not on. Observers attempt to merely observe the state of the world not force it to materialize.

**Comment by alexspeller** [comment]
@stefanpenner it is consumed in the component's template

**Comment by stefanpenner** [comment]
@alexspeller hmm, as my example (which consumes explicitly) works. I suspect something is funky when a template stream consumes it

@mmun thoughts?

**Comment by alexspeller** [comment]
@stefanpenner even weirder - here's the original example, where the property was explicitly consumed via `get`, which does not work in this case: http://emberjs.jsbin.com/zewoxagaki/2/edit

**Comment by alexspeller** [comment]
Also weird: changing your fiddle to explicitly consume the property in `didInsertElement` works fine:

http://emberjs.jsbin.com/sepenoqupu/2/edit

Changing @lolmaus's example to explicitly consume the property in `didInsertElement` does not work:

http://emberjs.jsbin.com/bipaxusaye/2/edit

**Comment by alexspeller** [comment]
@stefanpenner the difference with yours is that you use a *computed* property. Using a *non-computed* property causes your example to fail: 

http://emberjs.jsbin.com/lezetoyuka/2/edit

**Comment by mmun** [comment]
This is exactly the same "issue" as #10405.

TL;DR: If you call `set` on a property AND it is an object AND it is `===` to the existing value then observers will not be fired. [Here's the code](https://github.com/emberjs/ember.js/blob/1df0877edfccc3fb117f16cc88b071976d4ecd3b/packages/ember-metal/lib/property_set.js#L61).

Here is the sequence of events. I hope this helps you debug in the future.

- The helper call for `{{foo-bar myProperty=myProperty}}` executes.
  - The foo-bar component is created with a stream binding for `myProperty`.
    - The component begins initializing.
    - All stream bindings are connected.
      - The stream binding calls `set(component, "myProperty", get(controller, "myProperty))`. Note that this doesn't trigger observers because objects never fire observers during the init phase.
      - Set up observers on each side that sync the value back and forth between component and controller.
    - The component is finished initializing.
    - The stream binding is `get`s the value from the context (controller) and `set`
  - The component is finished instantiating.

- `controller.notifyPropertyChange("myProperty")` is called.
  - The `report` observer on the controller fires.
  - The `myProperty` stream binding observer fires.
    - The binding schedules to sync on the `sync` queue, but I will omit the details of this since it's not relevant.
    - The stream binding calls `set(component, "myProperty", get(controller, "myProperty))`. But the value hasn't changed! So the observers on the component are not notified as per the TL;DR above.

**Comment by alexspeller** [comment]
Thanks very much @mmun for the detailed explanation.

So is this really not a bug? It seems weird that `notifyPropertyChange` notifies changes to a property only in some circumstances but not in others, and that it works on a controller but not the component it's bound to. Unless it's a computed property. I still would call this a bug because whilst understanding the implementation helps to make sense of it, it seems very inconsistent behaviour

**Comment by OrKoN** [comment]
@mmun @alexspeller Do I understand it right that the `notifyPropertyChange` method works completely correct and the problem is that the value of myProperty does not actually change (because of the `===` check) and does not propagate to the component? And the right way to deal with it is to update the value of `myProperty` (and using `set` seems to be the only way?). And in such a case, the `notifyPropertyChange` call should not really needed.

Btw, it seems that it works for computed properties because computes properties return a new object each time:  [does not work here, because the object returned by a computed property is the same](http://emberjs.jsbin.com/kuyahusisa/3/edit?html,js,console,output)

**Comment by alexspeller** [comment]
That is really weird, because the computed property should be cached and should return the same object in the previous example.

**Comment by OrKoN** [comment]
@alexspeller a jsbin with both types of comp properties: http://emberjs.jsbin.com/kejotuweno/1/edit?html,js,console,output Yeah, I thought so too. A bug?

**Comment by alexspeller** [comment]
Oh, this explains it, `notifyPropertyChange` causes the computed property to actually *recalculate*, so it's not just notifying observers that it has changed but it's actually updating that property itself:

http://emberjs.jsbin.com/duxaxutuqe/2/edit

**Comment by mmun** [comment]
@alexspeller The `notifyPropertyChange` on the controller is working correctly. It is firing the observer on the controller as well as the observer that triggers the binding to synchronize.

Bindings are composed of two parts: the "from" path and the "to" path. In this case the "from" path points to a property on the controller and the "to" path is to a property on the component.

When the binding is created an observer is added on both the "from" and "to" paths. When `notifyPropertyChange` is called, it triggers the "from" observer which schedules a synchronization of the binding in the run loop (into the sync queue). Synchronization just means to call `set` on the _"to"_ side with the value of the "from" side. THIS is precisely where the issue lies. This `set` call becomes a no op if the value it is being set to is the same as it was previously.

This is the case here because the value is always that same POJO.

**Comment by mmun** [comment]
`notifyPropertyChange` just doesn't recompute anything. Specifically it:

- clears the cache if called on a key that is a CP
- recursively calls `notifyPropertyChange` if there are CPs that depend on this key
- notifies observers listening to this key

It doesn't automatically get recompute. Usually one of the notified observers will recompute the value.

**Comment by OrKoN** [comment]
Great, then it makes sense! Does not look like a bug to me now... To make it behave more straight-forward, an implementation of something like obj.equals(obj) for everything would be needed, I guess... 

I personally hit this issue only once for the all 8 months of using Ember.js and I think I used the workaround mentioned here. But anyway maybe this case needs to be emphasized in the Ember guide?

**Comment by alexspeller** [comment]
@mmun I adapted some of your wording here to improve the docs: https://github.com/emberjs/ember.js/pull/10664

**Comment by Glavin001** [comment]
My final workaround: http://stackoverflow.com/a/32411730/2578205

> This code below will allow your components to be passed data with different property keys on the controller. For instance, if the controller has a property data or wants to use the model from the router, the property key does not matter. The component does not need to have a fixed property key that is always used on the controller, such as "foo", instead it will dynamically find it.

```javascript
didInsertElement: function() {
    var controller = this.get('targetObject');
    // Find the key on the controller for the data passed to this component
    // See http://stackoverflow.com/a/9907509/2578205
    var propertyKey;
    var data = this.get('data');
    for ( var prop in controller ) {
        if ( controller.hasOwnProperty( prop ) ) {
             if ( controller[ prop ] === data ) {
               propertyKey = prop;
               break;
             }
        }
    }
    if (Ember.isEmpty(propertyKey)) {
      console.log('Could not find propertyKey', data);
    } else {
      console.log('Found key!', propertyKey, data);
      controller.addObserver(propertyKey, this, this.onDataChange);
    }
  }
```

---

I would really prefer that Ember support this internally between Controllers and Components.

**Comment by lolmaus** [comment]
@Glavin001 Can you please provide a JSBin demo with this workaround disabled and enabled?

**Comment by Glavin001** [comment]
@lolmaus here is my JSBin demo: http://emberjs.jsbin.com/nafapo/edit?console,output

**Comment by lolmaus** [comment]
@Glavin001 Awesome! It looks like it can be generalized into an addon.
